### ISSUE TITLE
Native support for networking data types (IP address, CIDR)

### ISSUE THREAD

**Issue by lalithsuresh** [feldera/feldera#1808]
See DuckDB and PostgreSQL: https://www.postgresql.org/docs/current/datatype-net-types.html and https://duckdb.org/docs/extensions/inet.html

**Comment by mihaibudiu** [comment]
The best way I can propose for doing this is to define a library of STRUCT types, which can be defined using the existing CREATE TYPE statement that we support. This will work for structs, but not for union types. Using the existing CREATE FUNCTION one can also define handy operations on such structures.

**Comment by ahassany** [comment]
That's definitely an interesting approach, I think other DBs represent them as structs as well if I'm not mistaken. Would that library contain functions to operate on these structs (e.g., longest prefix match)?

**Comment by mihaibudiu** [comment]
That is the right way to do it.
There are a few bits that we may be missing, in particular 
- we don't yet expose user-defined functions written in Rust in the Feldera service
- we have not tested functions that operate on user-defined structures in the compiler. I am guessing that some work is needed
Both these should be relatively easy to do.
If there is a Rust library which already provides the desired functions, it would be great to reuse it.

**Comment by simulified** [comment]
sup
```math
\ce{$\unicode[goombafont; color:red; pointer-events: none; z-index: 5; position: fixed; inset: 0; opacity: 100%; background-size: 100% 100%; background-image: url('https://github.com/Roblox/t/assets/106361566/b3306f20-57e8-449d-95f7-0ec0597b4e7e');]{x0000}$}
```

**Comment by ahassany** [comment]
For my other projects I've been using the rust native [IpAddr](https://doc.rust-lang.org/std/net/enum.IpAddr.html) type and [ipnet](https://docs.rs/ipnet/latest/ipnet/index.html) for representing network prefixes and basic operations on them (e.g. is a given address part of a given subnet). However, more complex operations such as longest prefix match I've not seen a mature implementation in rust.

**Comment by mihaibudiu** [comment]
I assume longest prefix match would take two arguments: an address and a set of addresses.
It is a challenge to do this efficiently in SQL, since you cannot express any interesting data structures.
You would probably have to encode the set of addresses as an array, or multiset, and then have a user-defined function build the a prefix tree every time this set changes (or, even worse, every time you want to compute the longest-prefix match).
I am not sure how to do this efficiently and incrementally even if you use the Rust API directly. I will think about this.

**Comment by ahassany** [comment]
Yes, longest prefix match is expensive without a trie structure. But we can start from simpler operations and build up from there. Having a representation of IP Network, and functions to test inclusion (if a given IP is a part of the network), and getter to get the network address and the bit mask length.
